shell/common/shell.cc

```diff
@@ -130,46 +133,67 @@ std::unique_ptr<Shell> Shell::CreateShellOnPlatformThread(
-        // from the platform.
-        auto animator = std::make_unique<Animator>(*shell, task_runners,
-                                                   std::move(vsync_waiter));
-
-        engine_promise.set_value(std::make_unique<Engine>(
-            *shell,                         //
-            dispatcher_maker,               //
-            *shell->GetDartVM(),            //
-            std::move(isolate_snapshot),    //
-            task_runners,                   //
-            window_data,                    //
-            shell->GetSettings(),           //
-            std::move(animator),            //
-            weak_io_manager_future.get(),   //
-            unref_queue_future.get(),       //
-            snapshot_delegate_future.get()  //
-            ));
-      }));
+      fml::MakeCopyable(
+          [engine_promise = std::move(engine_promise),
+           shell = shell.get(),  //
+           dispatcher_maker = std::move(dispatcher_maker),
+           window_data = std::move(window_data),
+           isolate_snapshot = std::move(isolate_snapshot),  //
+           vsync_waiter = std::move(vsync_waiter),          //
+           weak_io_manager_future = std::move(weak_io_manager_future),
+           snapshot_delegate_future = std::move(snapshot_delegate_future),
+           unref_queue_future = std::move(unref_queue_future)]() mutable {
+            TRACE_EVENT0("flutter", "ShellSetupUISubsystem");
+            const auto& task_runners = shell->GetTaskRunners();
+
+            // The animator is owned by the UI thread but it gets its vsync
+            // pulses from the platform.
+            auto animator = std::make_unique<Animator>(*shell, task_runners,
+                                                       std::move(vsync_waiter));
+
+            engine_promise.set_value(std::make_unique<Engine>(
+                *shell,                         //
+                dispatcher_maker,               //
+                *shell->GetDartVM(),            //
+                std::move(isolate_snapshot),    //
+                task_runners,                   //
+                window_data,                    //
+                shell->GetSettings(),           //
+                std::move(animator),            //
+                weak_io_manager_future.get(),   //
+                unref_queue_future.get(),       //
+                snapshot_delegate_future.get()  //
+                ));
+          }));
 
-  if (!shell->Setup(std::move(platform_view),  //
-                    engine_future.get(),       //
-                    rasterizer_future.get(),   //
-                    io_manager_future.get())   //
-  ) {
-    return nullptr;
-  }
+  // Creating engine takes noticeable time, get it to background waiting for the
+  // creation and then setup the shell.
+  auto result = std::async(
+      std::launch::async,
+      [shell = shell.get(), platform_view = std::move(platform_view),
```

Comment:
Sending a raw pointer `shell` to an async operation is not going to cut it.  You'll need a shared_ptr.

Reply:
I'm not quite sure whether I've caught you. What's the meaning of 'cut it'? Do you mean that `shell` might have been destructed while executing the async operation? In that case, shall I use `shell = shell->weak_factory_.GetWeakPtr()` instead?

Reply:
Sorry, "not going to cut it" is an english idiom that means "isn't going to be adequate".

Yes, the shell could be deleted before this operation or during this operation.  A weak pointer isn't going to be good enough.  In the case they are happening at the same time you'd need to block the deleting thread until this has completed.  The easiest solution is using a shared_ptr.